src/libpriv/rpmostree-core.c

```diff
@@ -1832,6 +1830,150 @@ run_posttrans_sync (int tmp_metadata_dfd,
   return TRUE;
 }
 
+static gboolean
+run_pre_sync (int tmp_metadata_dfd,
+              int rootfs_dfd,
+              DnfPackage *pkg,
+              GHashTable *ignore_scripts,
+              GCancellable *cancellable,
+              GError    **error)
+{
+  _cleanup_rpmheader_ Header hdr = NULL;
+
+  if (!get_package_metainfo (tmp_metadata_dfd, pkg, &hdr, NULL, error))
+    return FALSE;
+
+  if (!rpmostree_pre_run_sync (pkg, hdr, ignore_scripts,
+                               rootfs_dfd, cancellable, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static gboolean
+apply_rpmfi_overrides (int            tmp_metadata_dfd,
+                       int            tmprootfs_dfd,
+                       DnfPackage    *pkg,
+                       GHashTable    *passwdents,
+                       GHashTable    *groupents,
+                       GCancellable  *cancellable,
+                       GError       **error)
+{
+  int i;
+  _cleanup_rpmfi_ rpmfi fi = NULL;
+
+  if (!get_package_metainfo (tmp_metadata_dfd, pkg, NULL, &fi, error))
+    return FALSE;
+
+  while ((i = rpmfiNext (fi)) >= 0)
+    {
+      const char *fn = rpmfiFN (fi);
+      const char *user = rpmfiFUser (fi) ?: "root";
+      const char *group = rpmfiFGroup (fi) ?: "root";
+      const char *fcaps = rpmfiFCaps (fi) ?: '\0';
+      rpm_mode_t mode = rpmfiFMode (fi);
+      struct stat stbuf;
+      uid_t uid = 0;
+      gid_t gid = 0;
+
+      if (g_str_equal (user, "root") &&
+          g_str_equal (group, "root"))
+        continue;
+
+      if (!S_ISREG (mode) &&
+          !S_ISDIR (mode) &&
+          !S_ISLNK (mode))
+        continue;
+
+      g_assert (fn != NULL);
+      while (fn[0] == '/')
+        fn += 1;
+
+      /* /run and /var paths have already been tranlated to tmpfiles during
+       * unpacking */
+      if (g_str_has_prefix (fn, "run/") ||
+          g_str_has_prefix (fn, "var/"))
+        continue;
+
+      if (fstatat (tmprootfs_dfd, fn, &stbuf, AT_SYMLINK_NOFOLLOW) != 0)
+        {
+          glnx_set_prefix_error_from_errno (error, "%s", "fstatat");
+          return FALSE;
+        }
+
+      if ((S_IFMT & stbuf.st_mode) != (S_IFMT & mode))
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                       "Inconsistent file type between RPM and checkout "
+                       "for file '%s' in package '%s'", fn,
+                       dnf_package_get_name (pkg));
+          return FALSE;
+        }
+
+      if (S_ISREG (mode))
+        {
+          if (!break_single_hardlink_at (tmprootfs_dfd, fn, cancellable, error))
+            return FALSE;
+        }
```

Comment:
Hmm, yeah that makes sense. On my machine, the only such example I could find was:

```
$ ls -la /usr/share/bash-completion/completions/mock
-rw-r--r--. 1 root mock 5810 Sep 13 05:15 /usr/share/bash-completion/completions/mock
```

which looks extremely dubious.

I changed it to only apply overrides to files and directories. If you feel strongly about it, we can make it a hard error in the unpacker or so.